File: security_cam/lib/sensors/simulated_audio_source.dart
import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import '../core/audio_source.dart';
import '../core/models.dart';

enum AudioScene { silence, babyCry, glassBreak, bang }

class SimulatedAudioSource implements AudioSource {
  static const sampleRate = 16000;
  static const windowSamples = 15600;

  final StreamController<AudioWindow> _controller =
      StreamController<AudioWindow>.broadcast();
  Timer? _timer;
  AudioScene scene = AudioScene.babyCry;

  @override
  Stream<AudioWindow> get windows => _controller.stream;

  @override
  void start() {
    _timer ??= Timer.periodic(const Duration(seconds: 1), (_) {
      _controller.add(AudioWindow(
        timestamp: DateTime.now(),
        samples: generateWindow(scene),
        sampleRate: sampleRate,
      ));
    });
  }

  @override
  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  static Float32List generateWindow(AudioScene scene) {
    final rng = Random(42);
    final samples = Float32List(windowSamples);
    switch (scene) {
      case AudioScene.silence:
        for (var i = 0; i < windowSamples; i++) {
          samples[i] = (rng.nextDouble() - 0.5) * 0.02;
        }
      case AudioScene.babyCry:
        final freq = 250.0;
        for (var i = 0; i < windowSamples; i++) {
          final t = i / sampleRate;
          final mod = 0.7 + 0.3 * sin(2 * pi * 4 * t);
          samples[i] = 0.45 * mod * sin(2 * pi * freq * t);
        }
      case AudioScene.glassBreak:
        for (var i = 0; i < windowSamples; i++) {
          final envelope = exp(-1 * (i / windowSamples));
          samples[i] = 0.8 * envelope * (rng.nextDouble() * 2 - 1);
        }
      case AudioScene.bang:
        for (var i = 0; i < windowSamples; i++) {
          samples[i] = rng.nextDouble() * 2 - 1;
        }
    }
    return samples;
  }

  @override
  Future<void> dispose() async {
    stop();
    await _controller.close();
  }
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/sensors/SimulatedAudioSource.kt
package io.securitycam.level1.sensors

import kotlin.math.exp
import kotlin.math.sin
import kotlin.math.PI

enum class AudioScene { silence, babyCry, glassBreak, bang }

/**
 * Deterministic synthetic audio windows for tests and the simulated source
 * (port of `lib/sensors/simulated_audio_source.dart`). The streaming
 * AudioSource wrapper lands with the Phase 4 monitoring wiring.
 */
object SimulatedAudioSource {
    const val sampleRate = 16000
    const val windowSamples = 15600

    fun generateWindow(scene: AudioScene): FloatArray {
        val rng = java.util.Random(42)
        val samples = FloatArray(windowSamples)
        when (scene) {
            AudioScene.silence -> {
                for (i in 0 until windowSamples) {
                    samples[i] = ((rng.nextDouble() - 0.5) * 0.02).toFloat()
                }
            }
            AudioScene.babyCry -> {
                val freq = 250.0
                for (i in 0 until windowSamples) {
                    val t = i.toDouble() / sampleRate
                    val mod = 0.7 + 0.3 * sin(2 * PI * 4 * t)
                    samples[i] = (0.45 * mod * sin(2 * PI * freq * t)).toFloat()
                }
            }
            AudioScene.glassBreak -> {
                for (i in 0 until windowSamples) {
                    val envelope = exp(-1.0 * (i.toDouble() / windowSamples))
                    samples[i] = (0.8 * envelope * (rng.nextDouble() * 2 - 1)).toFloat()
                }
            }
            AudioScene.bang -> {
                for (i in 0 until windowSamples) {
                    samples[i] = (rng.nextDouble() * 2 - 1).toFloat()
                }
            }
        }
        return samples
    }
}